components/raftstore-v2/src/worker/cleanup/compact.rs

```diff
@@ -0,0 +1,365 @@
+                region_ids,
+                compact_threshold,
+            } => match collect_regions_to_compact(
+                &self.tablet_registry,
+                region_ids,
+                compact_threshold,
+                &self.logger,
+            ) {
+                Ok(mut region_ids) => {
+                    for region_id in region_ids.drain(..) {
+                        let Some(mut tablet_cache) = self.tablet_registry.get(region_id) else {continue};
+                        let Some(tablet) = tablet_cache.latest() else {continue};
+                        for cf in &cf_names {
+                            if let Err(e) =
+                                tablet.compact_range_cf(cf, None, None, false, 1 /* threads */)
+                            {
+                                error!(
+                                    self.logger,
+                                    "compact range failed";
+                                    "regionid" => region_id,
+                                    "cf" => cf,
+                                    "err" => %e,
+                                );
+                            }
+                        }
+                        info!(
+                            self.logger,
+                            "compaction range finished";
+                            "region_id" => region_id,
+                        );
+                        fail_point!("raftstore-v2::CheckAndCompact::AfterCompact");
+                    }
+                }
+                Err(e) => warn!(
+                    self.logger,
+                    "check ranges need reclaim failed"; "err" => %e
+                ),
+            },
+        }
+    }
+}
+
+fn need_compact(range_stats: &RangeStats, compact_threshold: &CompactThreshold) -> bool {
+    if range_stats.num_entries < range_stats.num_versions {
+        return false;
+    }
+
+    // We trigger region compaction when their are to many tombstones as well as
+    // redundant keys, both of which can severly impact scan operation:
+    let estimate_num_del = range_stats.num_entries - range_stats.num_versions;
+    let redundant_keys = range_stats.num_entries - range_stats.num_rows;
+    (redundant_keys >= compact_threshold.redundant_rows_threshold
+        && redundant_keys * 100
+            >= compact_threshold.redundant_rows_percent_threshold * range_stats.num_entries)
+        || (estimate_num_del >= compact_threshold.tombstones_num_threshold
+            && estimate_num_del * 100
+                >= compact_threshold.tombstones_percent_threshold * range_stats.num_entries)
+}
+
+fn collect_regions_to_compact<E: KvEngine>(
+    reg: &TabletRegistry<E>,
+    region_ids: Vec<u64>,
+    compact_threshold: CompactThreshold,
+    logger: &Logger,
+) -> Result<Vec<u64>, Error> {
+    fail_point!("on_collect_regions_to_compact");
+    info!(
+        logger,
+        "received compaction check";
+        "regions" => ?region_ids
```

Comment:
Could limit the number of printed region id? Otherwise it may overwhelm log.

Reply:
Modifying it to debug is also fine.